### ISSUE TITLE
qvm-backup-restore is slow

### ISSUE THREAD

**Issue by woju** [QubesOS/qubes-issues#2986]
#### Qubes OS version (e.g., `R3.2`): R4.0

`qvm-backup-restore` pipes internally whole backup tarfile at least three times (header, qubes.xml, data). This is painful on big backups and frustrating if something goes wrong at a later stage for large (~1TB) backups. (cf. bliviet)

Ideally reading should be streamlined to at most two passes: 1) header + hmac check; 2) the rest. Tarfile layout should explicitly support this, if it does not already.

Cc: @marmarek

**Comment by marmarek** [comment]
It does this only when you connect backup source directly to dom0, which you shouldn't do.

**Comment by marmarek** [comment]
And for the dom0 case, this is limitation of tar. If you find a way tell tar to stop reading the archive, when all the files mentioned on command line are found, that would greatly improve the situation. Otherwise, alternative, hacky, solution is to keep track of restored files in python and kill the tar process when it get all requested files - something I'd like to avoid.

As for your idea - it isn't that simple, because you know what files to expect (including qubes.xml + qubes.xml.hmac vs qubes.xml.enc) only after retrieving backup header.

**Comment by woju** [comment]
I did it from separate VM, granting it (unsuccessfully at first, that's how I know) necessary Admin API permissions. I felt it does three passes.

For some ideas how to do that, I'd rewrite it using `tarfile` and I'd forget about the part of spec that says the file that counts is the last one in the archive.

I probably know nothing about corner cases, but this doesn't take 2.5 hours:
```python
import tarfile
t = tarfile.open('qubes-backup', ignore_zeros=True)

ti = t.next()
assert ti.name == 'backup-header'
backup_header = t.extractfile(ti).read()

ti = t.next()
assert ti.name == 'backup-header.hmac'
backup_header_hmac = t.extractfile(ti).read()
```

**Comment by marmarek** [comment]
While this may work for backup header, loading actual backup data into RAM is *REALLY BAD IDEA* (you've mentioned 1TB, no?). And having two different methods for extracting and verifying backup header and backup data is asking for bug in either of them.

**Comment by woju** [comment]
Hmmm.

```
Process Process-2:2:
Traceback (most recent call last):
  ...
  File "/usr/lib/python3/dist-packages/qubesadmin/backup/restore.py", line 1698, in _handle_appmenus_list
    stdin=stream)
  ...
FileNotFoundError: [Errno 2] No such file or directory: 'qvm-appmenus'
qubesadmin.backup.extract: Error while processing ...
```

and the same for all vms.

Looks like another 6 hours till I get back.

**Comment by woju** [comment]
It's not into RAM. `t.extractfile` gives a file-like object.

**Comment by marmarek** [comment]
Running qvm-backup-restore in dedicated VM = paranoid restore mode. Not supporting appmenus is somehow intentional missing feature.

**Comment by woju** [comment]
Does it even work at all?

**Comment by marmarek** [comment]
You tell me...

**Comment by woju** [comment]
I tell you it doesn't. Just don't tell Joanna...

**Comment by marmarek** [comment]
> It's not into RAM. t.extractfile gives a file-like object.

Ah, so you're seeking solution for performance related issues, by passing the whole blob through python?

**Comment by woju** [comment]
No. I'm seeking solutions for performance-related issues by intentionally forgetting about relevant parts of tar file specification. If it takes python to misimplement it, so be it, if the alternative is to do it in C.

**Comment by marmarek** [comment]
In theory, `tar --seek` should also greatly improve performance (it should simply seek through all ignored files). But in also in theory it should be automatically enabled...

**Comment by woju** [comment]
No, it won't. According to python's `tarfile` module documentation, the file that counts under a given path is the last one in the archive (I think it this is that way to be able to do incremental backups on tape by simply appending new files). So you have to seek whole file anyway, especially with `-i`. GNU tar works with this assumption. I say we break that and hereby declare header and HMAC to be the first two files in the archive.

**Comment by marmarek** [comment]
But seeking through the whole archive (just reading file headers) also should be much faster, even for 1TB archive.
I'm checking `tar --occurrence` option right now, which should do exactly what we want.

**Comment by marmarek** [comment]
For simple archive it helps. If you're brave enough, try adding `--occurrence=1` to https://github.com/QubesOS/qubes-core-admin-client/blob/master/qubesadmin/backup/restore.py#L870

**Comment by woju** [comment]
Me? Not brave enough? Just hold me my beer...

**Comment by marmarek** [comment]
You should get backup summary in seconds.

**Comment by woju** [comment]
It got better!

**Comment by woju** [comment]
Hmmm. Manpage says default for that option is 1.

Maybe the truth is "avaiable in the texinfo format".

**Comment by marmarek** [comment]
That was also my understanding, but now I think `--occurrence[=N]` means "The default N is 1, if you specify --occurrence without N".

**Comment by woju** [comment]
Guys, I'm going to fix the issue about `--occurence`, but it leaves the bug around `qvm-appmenus`, so I'm going to file another ticket for that.
